File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/cache/operation/OutOfBandOperation.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.cache.operation

import ch.nevis.mobile.authentication.sdk.react.cache.state.NoOperation
import ch.nevis.mobile.authentication.sdk.react.cache.state.OperationState
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.userinteraction.onerror.DefaultOnErrorImpl
import ch.nevis.mobile.authentication.sdk.react.userinteraction.onsuccess.DefaultOnSuccessImpl
import ch.nevis.mobile.sdk.api.MobileAuthenticationClientError

data class OutOfBandOperation(
    override var state: OperationState,
    override var onSuccess: (Unit?) -> Unit,
    override var onError: (MobileAuthenticationClientError) -> Unit
) : Operation<Unit>() {
    constructor(operationId: String) : this(
        onSuccess = {
            DefaultOnSuccessImpl(operationId, ReactMethodName.OOB_OPERATION).onSuccess()
        },
        onError = {
            DefaultOnErrorImpl(operationId, ReactMethodName.OOB_OPERATION).onError(it)
        },
        state = NoOperation
    )

    override fun cancel() {
        state.cancel()
    }
}

File: ios/Cache/Operation/OutOfBandOperation.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

class OutOfBandOperation: Operation {
	// MARK: Properties

	var state: OperationState
	var onError: ErrorHandler

	// MARK: Initialization

	init(operationId: String) {
		self.state = NoOperationState()
		self.onError = { OnErrorImpl(operationId: operationId, method: .oobOperation).onError($0) }
	}

	// MARK: Operation

	func cancel() {
		state.cancel()
	}
}
